pkg/engine/internal/scheduler/scheduler.go

```diff
@@ -0,0 +1,718 @@
+	if task.status.State.Terminal() {
+		s.deleteTask(task)
+	}
+
+	if changed {
+		// Inform the owner about the change.
+		task.handler(ctx, task.inner, msg.Status)
+	}
+	return nil
+}
+
+func (s *Scheduler) handleStreamStatus(ctx context.Context, msg wire.StreamStatusMessage) error {
+	s.resourcesMut.Lock()
+	defer s.resourcesMut.Unlock()
+
+	stream, found := s.streams[msg.StreamID]
+	if !found {
+		return fmt.Errorf("stream %s not found", msg.StreamID)
+	}
+	return s.changeStreamState(ctx, stream, msg.State)
+}
+
+// changeStreamState updates the state of the target stream. changeStreamState
+// must be called with the resourcesMut lock held.
+func (s *Scheduler) changeStreamState(ctx context.Context, target *stream, newState workflow.StreamState) error {
+	changed, err := target.setState(newState)
+	if err != nil {
+		return err
+	} else if !changed {
+		return nil
+	}
+
+	// If we have a receiver, inform them about the change. This is a
+	// best-effort message so we don't need to wait for acknowledgement.
+	receiver, found := s.tasks[target.taskReceiver]
+	if found && receiver.owner != nil {
+		_ = receiver.owner.SendMessageAsync(ctx, wire.StreamStatusMessage{
+			StreamID: target.inner.ULID,
+			State:    newState,
+		})
+	}
+
+	// Inform the owner about the change.
+	target.handler(ctx, target.inner, newState)
+	return nil
+}
+
+// abortWorkerTasks immediately aborts all tasks assigned to the given worker.
+// abortWorkerTasks should only be used when the worker has disconnected.
+//
+// If the reason argument is nil, the tasks are cancelled. Otherwise, they are
+// marked as failed with the provided reason.
+func (s *Scheduler) abortWorkerTasks(ctx context.Context, worker *wire.Peer, reason error) {
+	s.resourcesMut.Lock()
+	defer s.resourcesMut.Unlock()
+
+	newStatus := workflow.TaskStatus{State: workflow.TaskStateCancelled}
+	if reason != nil {
+		newStatus = workflow.TaskStatus{
+			State: workflow.TaskStateFailed,
+			Error: reason,
+		}
+	}
+
+	for task := range s.workerTasks[worker] {
+		if changed, _ := task.setState(newStatus); !changed {
+			continue
+		}
+
+		s.deleteTask(task)
+
+		// We only need to inform the handler about the change. There's nothing
+		// to send to the owner of the task since worker has disconnected.
+		task.handler(ctx, task.inner, newStatus)
+	}
+
+	delete(s.workerTasks, worker)
+}
+
+func (s *Scheduler) runAssignLoop(ctx context.Context) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-s.assignSema:
+			s.assignTasks(ctx)
+		}
+	}
+}
+
+func (s *Scheduler) assignTasks(ctx context.Context) {
+	// We need to grab the lock on resources to prevent stream states from being
+	// modified while we're assigning the task.
+	//
+	// This prevents a race condition where a task owner misses a stream state
+	// change while we're assigning tasks at the same time as a state change.
+	//
+	// TODO(rfratto): Is there going to be too much overhead for locking this
+	// for this long?
```

Comment:
my only real concern reading through this is around locking, both this point about overhead, and also the various methods that must be called while a certain lock is held. I worry there's a bit of hidden state there, and I wonder if there's something we could do there? not a problem we need to solve now, since they're all private and only called here, and well documented, but it's a bit of a concern.